### DISCUSSION TITLE
Reference for the upstream patches

### DISCUSSION THREAD

**Discussion by sup-main** [open-quantum-safe/liboqs#1825] [General]
Hello,

I was referring the doc for ml-dsa(https://github.com/open-quantum-safe/liboqs/blob/main/docs/algorithms/sig/ml_dsa.yml). 
The source code seems to be a mix of code from standard branch(https://github.com/pq-crystals/dilithium/tree/standard) of pq-crystals/Dilithium and some patches(https://github.com/open-quantum-safe/liboqs/blob/main/scripts/copy_from_upstream/patches/pqcrystals-ml_dsa_ipd.patch)

Can I please know the source of the patches ? is it also from the pq-crystals team , do we a specific branch for it ?

**Comment by SWilson4** [comment]
Hi @iambatmanandjoker, `liboqs` developers create the patch files. Most of the time they are minor changes to ensure that upstream code integrates smoothly with the library (using our SHA3 code, for instance), but we sometimes patch errors in the upstream as well.

**Reply by sup-main** [reply]
Thanks @SWilson4 for the reply.
Sha3 code is where my doubts lie especially using shake256_inc_xx APIs.
Is that also developed by liboqs developer ? I see three version- simple,1600,avx2. How is it decided about which one to use ?

**Reply by baentsch** [reply]
The baseline answer is given [here](https://github.com/open-quantum-safe/liboqs/blob/main/CONFIGURE.md#oqs_use_openssl):

> These default choices have been made to optimize the default performance of all algorithms.

So for example, AVX2 code should be used on a platform indicating that support, OpenSSL SHA3 is only used when that is [configured](https://github.com/open-quantum-safe/liboqs/blob/main/CONFIGURE.md#oqs_use_openssl).
